core/src/main/java/io/questdb/cutlass/line/tcp/TableUpdateDetails.java

```diff
@@ -307,30 +310,73 @@ void clear() {
                 }
             }
             symbolCacheByColumnIndex.clear();
-            geoHashBitsSizeByColIdx.clear();
+            columnTypeMeta.clear();
+            columnTypeMeta.add(0);
         }
 
-        LowerCaseCharSequenceHashSet getAddedCols() {
-            return addedCols;
+        LowerCaseCharSequenceHashSet resetAddedCols() {
+            addedColsUtf16.clear();
+            return addedColsUtf16;
         }
 
-        int getColumnCount() {
-            return columnCount;
+        int getColumnIndex(DirectByteCharSequence colNameUtf8, boolean hasNonAsciiChars) {
+            int colIndex = columnIndexByNameUtf8.get(colNameUtf8);
+
+            // if the lookup in the cache is successful we can just return with the index
+            // if lookup was unsuccessful we have to check whether the column can be passed by name to the writer
+            if (colIndex < 0) {
+                final CharSequence colNameUtf16 = utf8ToUtf16(colNameUtf8, hasNonAsciiChars);
+                final int index = addedColsUtf16.keyIndex(colNameUtf16);
+
+                // non-negative index means this column is not in the set, meaning it has not been sent
+                // to the writer by name on this line before
+                // we can try to resolve column index using table reader
+                // if the column is in the set already (index is negative) the field has to be passed by name
+                if (index > -1) {
+                    colIndex = getColumnIndexFromReader(colNameUtf16.toString());
+
+                    // if we cannot not resolve column index even from the reader then this field
+                    // has not been created by the writer yet, it will be passed by name to the writer
+                    // if we resolve the index from the reader we will cache it
+                    if (colIndex > -1) {
+                        // keys of this map will be checked against DirectByteCharSequence when get() is called
+                        // DirectByteCharSequence.equals() compares chars created from each byte, basically it
+                        // assumes that each char is encoded on a single byte (ASCII)
+                        // we have to make sure that keys containing non-ASCII chars are converted into their
+                        // mangled utf8 representation before used as key (see more details in MangledUtf8Sink)
+                        columnIndexByNameUtf8.put(toMangledUtf8String(colNameUtf8.toString(), hasNonAsciiChars), colIndex);
```

Comment:
this seems complicated, all we need to do here is to store each byte from `colNameUtf8` as `char` into a StringSink (see the thread local sink)

Reply:
changed it